packages/core/src/common/message-rpc/rpc-message-encoder.ts

```diff
@@ -147,19 +250,216 @@ export interface ValueDecoder {
 }
 
 /**
- * Custom error thrown by the {@link RpcMessageEncoder} if an error occurred during the encoding and the
- * object could not be written to the given {@link WriteBuffer}
+ * Highly customizable {@link RpcMessageEncoder} implementation. Clients can register custom
+ * {@link ValueEncoder}s and {@link ValueEncoder}s to specially handling certain types of values.
+ * Objects are encoded recursively with the corresponding {@link ValueEncoder}.
+ * Compared to the default {@link MsgPackMessageEncoder} encoding times are higher (especially for large objects) so
+ * it's recommended to only use this encoder for edge cases where the capabilities of the default encoder don't suffice.
  */
-export class EncodingError extends Error {
-    constructor(msg: string) {
-        super(msg);
+export class RecursiveMessageEncoder implements RpcMessageEncoder {
```

Comment:
For  now I kept the "old" encoder as it offers complete flexibility in terms of encoding/decoding.
We have to decide whether we want to keep it as an alternative for adopters or if the msgpack based encoder suffices

Reply:
I'm generally always in favor of flexibility for users, however in this case I believe that's a bit of an overkill. Especially when it comes to the messaging API, I would try to keep the API surface as small as possible given it's overall complexity. I believe the new encoder already covers our bases well enough.